File: server/scheduler/briefing.js
// Composes the morning briefing: gathers real facts in code first — never
// invented — then asks the currently-selected model to narrate them into a
// short, spoken-style passage, under the same "phrase it yourself, don't
// invent data" guardrail every other tool result already follows (see
// prompt.js). Runs in its own conversation session so a briefing never
// mixes into whatever the user was chatting about.

import { listTasks } from './task-store.js';
import { listEntries as listProfileEntries } from '../profile.js';
import { invoke } from '../capabilities.js';
import { runTurn, resetConversation } from '../models/runner.js';
import { getBriefingConfig, setBriefingConfig } from './briefing-config.js';
import { toolNamesForConnector } from '../connectors/index.js';
import { listUserSkills } from '../skills/store/skill-files.js';

// Re-exported so existing importers (server.js, scheduler.js) don't need to
// care that these live in a separate, dependency-free file — see
// briefing-config.js's header comment for why the split exists.
export { getBriefingConfig, setBriefingConfig };

function timeGreeting() {
  const h = new Date().getHours();
  if (h < 12) return 'Good morning';
  if (h < 18) return 'Good afternoon';
  return 'Good evening';
}

/** Runs one built-in ability directly (not through a model) and returns its data, or null if it's off, failed, or no longer exists — never invented around. */
async function gatherOne(skillName, args, label) {
  const result = await invoke(skillName, args, { autoConfirm: true, source: 'briefing' });
  // NOT `!result.ok` — get_time returns no `ok` field on success at all,
  // which would make a bare `!result.ok` check silently drop it.
  if (!result || result.ok === false) {
    console.warn(`[briefing] "${skillName}" didn't work:`, result?.error);
    return null;
  }
  const { ok, needs_confirmation, confirm_token, ui_action, error, ...data } = result;
  return { label, data };
}

/**
 * Gathers weather and headlines — fixed, always-available native abilities,
 * not a user-managed list (see briefing-config.js's header comment for why
 * that shape was reverted). No UI controls either; `config.weatherPlace` is
 * set conversationally via configure_briefing.js.
 */
async function gatherLiveInfo(config) {
  const out = [];
  if (config.weatherPlace) {
    const weather = await gatherOne('get_weather', { place: config.weatherPlace }, `Weather in ${config.weatherPlace}`);
    if (weather) out.push(weather);
  }
  if (config.headlines) {
    const headlines = await gatherOne('get_headlines', { count: 4 }, 'News headlines');
    if (headlines) out.push(headlines);
  }
  return out;
}

/** A plain object's data as a short "key: value, key2: value2" string for the narration prompt — arrays joined, empties dropped. */
function formatSourceData(data) {
  return Object.entries(data)
    .filter(([, v]) => v !== undefined && v !== null && v !== '')
    .map(([k, v]) => `${k}: ${Array.isArray(v) ? v.join('; ') : typeof v === 'object' ? JSON.stringify(v) : v}`)
    .join(', ');
}

async function gatherFacts(config) {
  const facts = {};
  const now = new Date();

  if (config.sections.greeting) facts.greeting = timeGreeting();

  if (config.sections.dateTime) {
    facts.date = now.toLocaleDateString('en-US', { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' });
    facts.time = now.toLocaleTimeString('en-US', { hour: 'numeric', minute: '2-digit' });
  }

  if (config.sections.tasks) {
    facts.upcomingTasks = listTasks()
      .filter((t) => t.enabled && t.nextRunAt)
      .sort((a, b) => new Date(a.nextRunAt) - new Date(b.nextRunAt))
      .slice(0, 5)
      .map(
        (t) =>
          `${t.title} (${new Date(t.nextRunAt).toLocaleString('en-US', {
            weekday: 'short',
            hour: 'numeric',
            minute: '2-digit',
          })})`
      );
  }

  facts.liveInfo = await gatherLiveInfo(config);

  if (config.sections.goals) {
    const entries = listProfileEntries();
    if (entries.length) facts.goalsAndNotes = entries.slice(-8).map((e) => e.text);
  }

  if (config.sections.custom && config.customText) {
    facts.custom = config.customText;
  }

  return facts;
}

/** Compares what's enabled against what Jarvis knows, for gentle "want me to add..." suggestions. */
function suggestAdditions(config, facts) {
  const suggestions = [];
  if (!config.weatherPlace) suggestions.push("adding today's weather, if you tell me your city");
  if (!config.headlines) suggestions.push('including a few news headlines each morning');
  if (facts.goalsAndNotes?.length && !config.sections.focus) {
    suggestions.push('suggesting a focus for the day, based on what you\'ve told me about your goals');
  }
  return suggestions;
}

function factsToPrompt(facts, suggestions, config, usingConnectors) {
  const lines = [
    'Compose the morning briefing now, in your own words, as one short spoken passage — a few ' +
      'sentences, not a bulleted list. Use ONLY the facts below; never invent or guess anything not given here.',
    '',
  ];
  if (usingConnectors) {
    lines.push(
      'The user has also explicitly chosen to let this briefing use one or more of their connected apps — ' +
        'you have real tools available for exactly those connected apps, and no others. If checking one of them would ' +
        'genuinely add something useful to this briefing, call it; otherwise skip it rather than forcing it in. ' +
        'Same rule as everything else here: say only what a tool result actually returned, never a guess about ' +
        'what it might say.',
      ''
    );
  }
  if (facts.greeting) lines.push(`Greeting: ${facts.greeting}`);
  if (facts.date) lines.push(`Today: ${facts.date}, ${facts.time}`);
  if (facts.upcomingTasks?.length) lines.push(`Upcoming scheduled items: ${facts.upcomingTasks.join('; ')}`);
  for (const source of facts.liveInfo || []) {
    const formatted = formatSourceData(source.data);
    if (formatted) lines.push(`${source.label}: ${formatted}`);
  }
  if (facts.goalsAndNotes?.length) {
    lines.push(`Things the user has told Jarvis about themselves and their goals: ${facts.goalsAndNotes.join('; ')}`);
  }
  if (facts.custom) lines.push(`Custom note to include: ${facts.custom}`);
  // Was unconditional — the "Suggested focus for the day" checkbox
  // (briefing.js's SECTION_LABELS) toggled nothing at all, since this never
  // consulted config.sections.focus, only whether tasks/goals existed.
  if (config.sections.focus && (facts.upcomingTasks?.length || facts.goalsAndNotes?.length)) {
    lines.push(
      'If it genuinely fits, end with one brief suggested focus for the day grounded in the facts above — skip it rather than force one.'
    );
  }
  if (suggestions.length) {
    lines.push(
      '',
      `After the briefing itself, on a new line, briefly suggest ONE of these possible additions the user might want: ${suggestions.join(' / ')}`
    );
  }
  return lines.join('\n');
}

/**
 * Real tool names for every connector id the user explicitly selected for
 * this briefing (config.connectors — see briefing-config.js) — empty by
 * default, nothing automatic. A stale id (a since-removed connector)
 * resolves to no names via toolNamesForConnector() rather than throwing, so
 * one bad saved id degrades silently instead of breaking the briefing.
 */
function connectorToolNames(config) {
  const names = [];
  for (const id of config.connectors || []) names.push(...toolNamesForConnector(id));
  return names;
}

/** Gathers facts and has the model narrate them. `modelId` optionally pins one specific model (falls back through the usual auto-ranked chain if it breaks). Returns {ok, text, facts, modelId, switchedFrom, switchReason}. */
export async function composeBriefing({ modelId } = {}) {
  const config = getBriefingConfig();
  const facts = await gatherFacts(config);
  const suggestions = suggestAdditions(config, facts);

  // Everything above still gathers in code and narrates only what it found
  // — the "never invent" guarantee this file exists to enforce. Selected
  // connectors are the one deliberate, explicit exception: the user picked
  // exactly these, so the turn is allowed to actually call their tools
  // too, on top of narrating the code-gathered facts above. No connectors
  // selected (the default) keeps today's exact behavior — a pure
  // narration-only turn with no tool access at all.
  const connectorTools = connectorToolNames(config);
  const usingConnectors = connectorTools.length > 0;
  // A Skill is the user's own instructions (a house style for the briefing itself, say) —
  // always offered here too, same as everywhere else a turn runs, but deliberately NOT
  // widened to the full core built-in set the way scheduler.js's prompt-action turn is:
  // a briefing stays a narrate-code-gathered-facts turn, and a Skill can't invent data any
  // more than the model itself can, so this doesn't weaken the guarantee above.
  const skillNames = listUserSkills().filter((s) => s.enabled).map((s) => s.name);
  const allowedTools = [...connectorTools, ...skillNames];
  const usingTools = allowedTools.length > 0;
  const prompt = factsToPrompt(facts, suggestions, config, usingConnectors);

  const sessionId = `briefing:${Date.now()}`;
  let text = '';
  let ok = true;
  let usedModelId = null;
  let switchedFrom = null;
  let switchReason = null;

  // addressed:true overrides the background gate specifically for the personality
  // framework and jobsSection() (see prompt.js's systemInstructionParts()) — a
  // briefing IS spoken to the owner, unlike a scheduled task's own prompt-action
  // turn or a Job worker's turn, even though it shares their background:true.
  const turnOpts = usingTools
    ? { background: true, addressed: true, source: 'text', modelId, autoConfirm: true, allowedTools }
    : { background: true, addressed: true, source: 'text', modelId, autoConfirm: true, noTools: true };

  for await (const ev of runTurn(sessionId, prompt, turnOpts)) {
    if (ev.type === 'model_switch' && !switchedFrom) {
      switchedFrom = ev.from;
      switchReason = ev.reason;
    }
    if (ev.type === 'done') {
      text = ev.text;
      usedModelId = ev.modelId;
    }
    if (ev.type === 'paused') {
      ok = false;
      text = ev.reason;
    }
  }
  resetConversation(sessionId);

  return { ok, text, facts, modelId: usedModelId, switchedFrom, switchReason };
}

File: backend/jarvis/scheduler/briefing.py
"""The morning briefing: facts gathered in code, narrated by the model.

**The guarantee this file exists to enforce: the model narrates, it never
supplies.** Every fact — the date, what is scheduled today, the weather, the
headlines, what the user has said matters to them — is gathered here, in code,
before any model call, and the prompt says plainly that anything not listed is
not known. A briefing that invents a meeting is worse than no briefing, because
it is indistinguishable from a real one.

That is also why the narration turn is TOOLLESS. Nothing it says can come from
anywhere except the facts handed to it.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

from ..gateway.client import NoModelAvailable, ask
from ..gateway.routing import Task as RoutingTask
from ..memory import store as memory_store
from .briefing_config import get_config
from .recurrence import describe
from .task_store import list_tasks

logger = logging.getLogger(__name__)


@dataclass
class Briefing:
    ok: bool
    text: str = ""
    facts: dict[str, Any] = field(default_factory=dict)
    model_id: str | None = None
    error: str | None = None


def gather_facts(config: dict[str, Any] | None = None,
                 now: datetime | None = None) -> dict[str, Any]:
    """Everything the briefing is allowed to mention. No model call happens here."""
    config = config or get_config()
    sections = config.get("sections") or {}
    now = now or datetime.now()
    facts: dict[str, Any] = {}

    if sections.get("dateTime"):
        hour = now.hour % 12 or 12
        facts["now"] = {
            "date": f"{now:%A, %B} {now.day}, {now.year}",
            "time": f"{hour}:{now.minute:02d} {'AM' if now.hour < 12 else 'PM'}",
            "partOfDay": ("morning" if now.hour < 12 else
                          "afternoon" if now.hour < 18 else "evening"),
        }

    if sections.get("tasks"):
        facts["tasks"] = [
            {"title": t["title"], "when": describe(t.get("recurrence")),
             "nextRunAt": t.get("nextRunAt")}
            for t in list_tasks() if t.get("enabled")
        ]

    if sections.get("goals"):
        # What the user has actually said matters, not an inference about it.
        facts["remembered"] = [m["text"] for m in memory_store.list_memories()[:8]]

    place = (config.get("weatherPlace") or "").strip()
    if place:
        from ..tools.get_weather import _run as weather

        result = weather(place=place)
        # Recorded either way: "I couldn't get the weather" is itself a fact the
        # briefing may state, and is far better than silence the user reads as
        # "there is nothing to say".
        facts["weather"] = result if result.get("ok") else {"ok": False,
                                                            "error": result.get("error")}

    if config.get("headlines"):
        from ..tools.get_headlines import _run as headlines

        result = headlines(count=4)
        facts["headlines"] = result.get("headlines") if result.get("ok") else []

    if sections.get("custom") and (config.get("customText") or "").strip():
        facts["custom"] = config["customText"].strip()

    return facts


def facts_to_prompt(facts: dict[str, Any], config: dict[str, Any],
                    using_connectors: bool = False) -> str:
    lines = [
        "Give the user their briefing, out loud, in your own voice.",
        "",
        "These are the ONLY facts you have. Anything not listed here, you do not "
        "know — say nothing about it, and never fill a gap with something "
        "plausible. If a section came back empty, either say so briefly or leave "
        "it out; do not invent an entry to fill the space.",
        "",
    ]
    if using_connectors:
        # The wording above would be false with tools on the table, and a stated
        # rule the model can see is not true is worse than no rule: it teaches
        # that the rules here are approximate.
        lines += [
            "They have also chosen to let this briefing use some of their connected "
            "apps. You have real tools for exactly those apps and no others. If "
            "checking one would genuinely add something, call it; otherwise skip it "
            "rather than forcing it in. The same rule holds either way: say only "
            "what a tool actually returned, never a guess about what it might say.",
            "",
        ]
    if facts.get("now"):
        lines.append(f"Now: {facts['now']['time']} on {facts['now']['date']} "
                     f"({facts['now']['partOfDay']}).")
    if "tasks" in facts:
        if facts["tasks"]:
            lines.append("Scheduled:")
            lines += [f"- {t['title']} ({t['when']})" for t in facts["tasks"]]
        else:
            lines.append("Scheduled: nothing.")
    if facts.get("weather"):
        weather = facts["weather"]
        lines.append(f"Weather: {weather['temperatureF']}F and {weather['condition']} "
                     f"in {weather['place']}." if weather.get("ok")
                     else f"Weather: unavailable ({weather.get('error')}).")
    if facts.get("headlines"):
        lines.append("Headlines:")
        lines += [f"- {h}" for h in facts["headlines"]]
    if facts.get("remembered"):
        lines.append("Things you remember about them (context, not items to read out):")
        lines += [f"- {m}" for m in facts["remembered"]]
    if facts.get("custom"):
        lines.append(f"They also asked you to include: {facts['custom']}")

    if not (config.get("sections") or {}).get("greeting"):
        lines += ["", "Skip any greeting; start straight with the content."]
    return "\n".join(lines)


#: What a briefing says out loud, whichever path composes it.
NARRATOR = ("You are giving the user their briefing out loud. You narrate only "
            "the facts you are given, in plain spoken sentences, warmly and "
            "briefly. You never invent an item, a time, or a detail.")


def connector_tool_names(config: dict[str, Any]) -> list[str]:
    """Real tool names for every connector the person picked for this briefing.

    Resolved HERE, at compose time, every time. A connector's tool list changes
    when it is reconnected, so a saved name would go stale silently; a saved id
    that no longer resolves simply contributes nothing.
    """
    from ..connectors.capabilities import tool_names_for

    names: list[str] = []
    for connector_id in config.get("connectors") or []:
        if connector_id:
            names.extend(tool_names_for(connector_id))
    return names


def compose_briefing(now: datetime | None = None) -> Briefing:
    """Gather the facts in code, then have a model narrate exactly those.

    **Everything above is gathered in code and only narrated**, which is the
    guarantee this file exists to keep: a briefing cannot invent an item because
    the model is never in a position to fetch one.

    **Chosen connectors are the one deliberate exception**, and it is explicit:
    the person picked exactly these, so the turn is allowed to actually call
    their tools on top of narrating what was gathered. With none picked — the
    default — this stays a narration-only turn with no tool access at all, which
    is the cheaper path as well as the stricter one.
    """
    config = get_config()
    facts = gather_facts(config, now)
    tools = connector_tool_names(config)
    prompt = facts_to_prompt(facts, config, using_connectors=bool(tools))

    if not tools:
        try:
            answer = ask(
                prompt, system=NARRATOR,
                # Nobody is waiting in real time, so cost matters more than latency.
                task=RoutingTask(text="briefing", background=True, needs_tools=False),
            )
        except NoModelAvailable as err:
            return Briefing(ok=False, facts=facts, error=str(err))
        return Briefing(ok=True, text=answer.text.strip(), facts=facts,
                        model_id=answer.model_id)

    return _compose_with_connectors(prompt, facts, tools)


def _compose_with_connectors(prompt: str, facts: dict[str, Any],
                             tools: list[str]) -> Briefing:
    """The tool-using path: a real turn, restricted to exactly what was chosen.

    Its own ephemeral session, never bound to chat history, so a briefing's
    working turns cannot appear in the conversation list. `addressed` is what
    separates this from a scheduled task's or a background job's turn, which
    also run in the background: those are not spoken to anyone, and this is.
    """
    import uuid

    from ..assembly import get_orchestrator
    from ..orchestrator import Done, Failed, TurnRequest
    from ..policy import Autonomy, Surface

    request = TurnRequest(
        text=prompt,
        session_id=f"briefing:{uuid.uuid4().hex[:8]}",
        surface=Surface.SCHEDULED,
        # Pre-consent covers ordinary work. A HIGH-risk action still parks for a
        # person — the owner's rule, enforced by the policy, not here.
        autonomy=Autonomy.PRE_CONSENTED,
        turn_id=uuid.uuid4().hex,
        allowed_names=frozenset(tools),
    )

    text, failure, model_id = "", None, None
    for event in get_orchestrator().run_turn(request):
        if isinstance(event, Done):
            text, model_id = event.text, event.model_id
        elif isinstance(event, Failed):
            failure = event
    if failure is not None:
        return Briefing(ok=False, text=text, facts=facts, error=failure.error)
    return Briefing(ok=True, text=text.strip(), facts=facts, model_id=model_id)
